# Welcome to lab_populations! 👥🌎🌍🌏👥

In lecture, you have been learning about both sampling and inference. This is the idea that we can calculate statistics from a random sample of the population and use those statistics to estimate what we would get if we asked every single person in the population a question. 

The goal of this lab is to gain a more intuitive understanding of what inference is. We will explore sampling from a population and the meaning behind confidence intervals, error, and the Central Limit Theorem (CLT).

<hr>

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab TA, the CAs, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

In [ ]:
# Meet your CAs and TA if you haven't already!
# First name is enough, we'll know who they are! :)
ta_name = ""
ca1_name = ""
ca2_name = ""
ca3_name = ""

# Work with your group again this week! 
# QOTD to Ask Your Group: "Did you see the eclipse?  Partial or full?  How was it?"
partner1_name = ""
partner1_netid = ""
partner1_eclipse = ""

partner2_name = ""
partner2_netid = ""
partner2_eclipse = ""

partner3_name = ""
partner3_netid = ""
partner3_eclipse = ""

<hr style="color: #DD3403;">

## Part 1: Sampling the Population

The `DISCOVERY_populations` library is included with this lab and contains a **very large** population (over 100,000 students) of current and former University of Illinois students.  We have simulated their answers to three questions:

1. Do you support the Kingfisher as the new Illinois mascot?
2. Do you follow @datascienceduo on Instagram?
3. Are you a Data Science major?

Right now, **we do NOT know the answers for the entire population and there is NO WAY to ask everyone**. Instead, we can only ask a sample of students and get answers for that sample. Run the following code to import the `DISCOVERY_populations` library and retrieve the sample:

In [12]:
import DISCOVERY_populations
from scipy.stats import norm
sample = DISCOVERY_populations.getSample()
sample.head()

DSmajor  FollowsDuo  ProKingfisher
119598        0           0              0
180344        1           1              1
157453        0           0              1
39944         1           1              1
147907        1           1              1

### Puzzle 1.1: Statistics about the Sample

You have received a **random sample** from the population and it looks like it has three columns: `DSmajor`, `FollowsDuo`, and `ProKingfisher`. Using the `len` function, create a variable `n` that stores the number of people in your sample:

In [13]:
n = len(sample)
n

44

We'll first focus on people who follow @datascienceduo -- the people who follow the DUO are coded with a `1` in the sample and the people who do not follow the DUO are coded with a `0`.  

In your sample, how many people follow the DUO?

In [14]:
followers = len(sample[sample["FollowsDuo"] == 1])
followers

23

followers = 
followers

In [15]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert("sample" in vars()), "Check to make sure you have the variable `sample`."
assert(len(sample) == n), "Check to make sure `n` stores the number of observations in your sample."
assert(followers == sum(sample.FollowsDuo)), "Check to make sure `followers` stores the number of people following @datascienceduo."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.2: Finding the 95% Confidence Interval for the Percentage of DUO followers

We want to estimate what percentage of the population follows @datascienceduo. To do that, we need to use the confidence interval formula you learned in lecture. 

$$ CI = {SamplePercent} \pm ({z} \times {SampleStandardError})$$

Let's work on finding all three of the components we need: `samplePercent`, `z`, and `sampleSE`. For this entire puzzle, make sure your percentages (samplePercent and sampleSE) are in percent form and not decimal form. In other words, they should be numbers between 0% and 100%.


#### Puzzle 1.2(a): Finding `samplePercent`

 Using the `FollowsDuo` column, store the **percentage of the sample that follow the DUO** in the variable `samplePercent`:
 
 *Note: Since the `FollowsDuo` column is encoded so that a `0` is a non-follower and a `1` is a follower, the mean of the column will be a proportion (decimal), but we want to find a **percentage** so make sure to convert your answer to be between 0 and 100 percent.*

In [16]:
samplePercent = (followers / n) * 100
samplePercent

52.27272727272727

In [17]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
import math
F = sample.FollowsDuo
assert(math.isclose(samplePercent, F.sum()/n*100)), "Check your `samplePercent`."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


#### Puzzle 1.2(b): Finding `z`

We want to find the range where we are 95% sure that the true percentage of people who follow the DUO is within that range. Find the z-score we need to use to create a 95% CI:

*Hint: Because the sample size is greater than 30 and the sample was randomly selected from the population, we can use the standard normal curve to find the z-score when creating our 95% CI.*

In [18]:
z = norm.ppf(0.95 + 0.025) 
z

1.959963984540054

In [19]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(abs(z) + abs(z)**abs(z), 5.69931068079139)), "Check your `z`."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")


🎉 All tests passed! 🎉


#### Puzzle 1.2(c): Finding `sampleSE`

Finally, we need to find the standard error of the sample as a **percentage**.

Remember: $SE_{\%} = \frac{SD}{\sqrt{n}} * 100$, where $SE$ is standard error, $SD$ is standard deviation, and $n$ is the sample size. 

In [20]:
sampleSE = sample["FollowsDuo"].std() / (n**0.5) * 100
sampleSE

7.617047451458002

In [21]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
assert(math.isclose(sampleSE, (n / F.var())**-0.5 * 100)), "Check your `sampleSE`."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 1.3: Finding the Confidence Interval

The formula for the confidence interval has both a "lower bound" (when you subtract the margin of error from the sample average) and an "upper bound" (when you add the margin of error to the sample average). Recall the formula you learned in lecture:

$$ CI = {SamplePercent} \pm ({z} \times {SampleStandardError})$$

Using the variables you just calculated in the previous section, find the `lower_bound_CI` and `upper_bound_CI` of your confidence interval:

In [22]:
lower_bound_CI = samplePercent - (sampleSE * z)
lower_bound_CI

37.34358859933698

In [23]:
upper_bound_CI = samplePercent + (sampleSE * z)
upper_bound_CI

67.20186594611756

Putting it all together, run the following code that will write out your full confidence interval interpretation:

In [24]:
print(f"Based on the sample, we are 95% confident that the true percentage of followers of @datascienceduo in the full population is between:\n   {round(lower_bound_CI, 2)}% - {round(upper_bound_CI, 2)}%")

Based on the sample, we are 95% confident that the true percentage of followers of @datascienceduo in the full population is between:
   37.34% - 67.2%


### Reflections

**Q1**: Talk to your group members and share your confidence intervals.
- (a): What is the confidence interval of another group member's sample?
- (b): Is it the same or different?
- (c): Why is it okay that it's the same or different?

*(✏️ The confidence interval was also 95%, but they had a smaller interval than mine.It is okay to have different answers because the sample is different. ✏️)*

**Q2**: What is the margin of error (the numeric value) of your confidence interval?  *You may need to do a bit of calculation.*

In [27]:
MOE = z * ((sample["FollowsDuo"].std()**2 / n)**0.5)
MOE

0.14929138673390288

*(✏️ MOE is around 0.149. ✏️)*

**Q3**: If the complete population is 1,000,000 people, we can be 95% certain **at least how many people are following the DUO**?  First, in English (not math equations), explain in at least one sentence how you will calculate this result.  Then, calculate it and include your answer below.

*(✏️ We can to the confidence interval for 95% with the population data or sample size of at least 30 people and find the lower bound of the interval. ✏️)*

**Q4**: Finally, can we accurately predict if at least 50% of the population follow @datascienceduo given your confidence interval?

*(✏️ No, since the interval goes below 50%, we can't conclude that at least 50% follow the page. ✏️)*

<hr style="color: #DD3403;">

## Part 2: Towards a Smaller Margin of Error

The number of followers of @datascienceduo is fun, but the large margin of error you had is a little alarming.  For really important issues, we want a smaller margin of error in our sample.

**Q5**: What are at least **TWO** things we can do as a data scientist to reduce the margin of error?

*(✏️ We can increase the sample size or reduce standard deviation. ✏️)*

### Part 2.1: An Expensive Sample

Taking a large sample requires surveying more people and getting more responses, which is almost always more expensive.  In the `DISCOVERY_populations` library you imported in Part 1, we have a second function: `getExpensiveSample()`.

The following code gets a larger and more expensive sample and stores it in `sample2`:

In [28]:
sample2 = DISCOVERY_populations.getExpensiveSample()
sample2

DSmajor  FollowsDuo  ProKingfisher
3351          0           1              1
122524        1           1              1
174931        1           1              1
12986         0           0              0
194787        0           0              0
...         ...         ...            ...
50722         1           1              1
162147        1           1              1
84173         0           0              0
105439        1           1              1
174412        1           1              1

[1006 rows x 3 columns]

### Part 2.2: Finding the Confidence Interval for Kingfisher Support

Find the lower and upper bounds for the 95% CI for the support of the Kingfisher mascot, storing them in `kingfisher_CI_lower` and `kingfisher_CI_upper`.  We provided individual cells for each stage of the computation, and you should make sure your answer is reasonable at each step. We also want your answers as percentages between 0 and 100 percent.

Make sure you're using `sample2` since you have the better, more expensive sample now! :)

In [31]:
# Step 1: Find the samplePercent:
samplePercent2 = (len(sample2[sample2["ProKingfisher"] == 1]) / len(sample2)) * 100
samplePercent2

66.8986083499006

In [32]:
# Step 2: Find the z-score for the 95% CI:
z = norm.ppf(0.95 + 0.025) 
z

1.959963984540054

In [33]:
# Step 3: Find the sampleSE:
sampleSE2 = sample2["ProKingfisher"].std() / (len(sample2)**0.5) * 100
sampleSE2

1.4843921099749522

In [34]:
# Step 4: Find the margin of error:
MOE2 = z * ((sample2["ProKingfisher"].std()**2 / len(sample2))**0.5)
MOE2

0.029093550744863254

In [38]:
# Find the lower bound of the CI:
kingfisher_CI_lower = samplePercent2 - (sampleSE2 * z)
kingfisher_CI_lower

63.98925327541427

In [39]:
# Find the upper bound of the CI:
kingfisher_CI_upper = samplePercent2 + (sampleSE2 * z)
kingfisher_CI_upper

69.80796342438693

In [40]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any errors, you PASSED all test cases!
# - If this cell results in any errors, check your previous cell, make changes, and RE-RUN your code and then this cell.
import math
from scipy.stats import norm

F = sample2.ProKingfisher
N = norm(F.mean(), F.std() / (len(F)**0.5))
low, high = N.interval(0.95)
assert(kingfisher_CI_upper > kingfisher_CI_lower), "The upper bound must be larger than the lower bound."
assert( math.isclose(kingfisher_CI_lower, low * 100) ), "Check your `kingfisher_CI_lower` calculation"
assert( math.isclose(kingfisher_CI_upper, high * 100) ), "Check your `kingfisher_CI_upper` calculation"

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Part 2.3: Reflections

**Q6**: Write out the interpretation of your confidence interval in a complete sentence.

*(✏️ We are 95% confidence that the average number of supporters for the prokingfisher mascot is between 63.98925327541427% to 69.80796342438693%. ✏️)*

**Q7**: If the whole population voted on if the next mascot should be the Kingfisher, how confident are you that the resolution will pass (that is, receive at least 50% of the vote)? Explain in at least one complete sentence how the data analysis you did backs up your confidence.

*(✏️ We are at least 95% confident that the next vote will pass because there is at most a 5% chance that the mean will fall outside the interval and be lower than 50%. ✏️)*

**Q8**: Is the confidence interval of your larger (and more expensive) sample larger (wider) or smaller (narrower) than the first sample?  Explain in at least one complete sentence.

*(✏️ My original confidence interval was smaller under the influence of randomness sampling, but according to theory, the interval should become narrower with the increase in sample size. ✏️)*

<hr style="color: #DD3403;">

## Part 3: The Election is Here!

The polling is complete and the election day is here!  Run the following code to find your election-day results:

In [41]:
DISCOVERY_populations.electionDay()

The election was held and 19% of the population voted.

== Kingfisher Support ==
SUPPORT KINGFISHER: 26567 69.91%
OPPOSE KINGFISHER : 11433 30.09%

== Follows @datascienceduo ==
FOLLOWS DUO    : 19774 52.04%
DOES NOT FOLLOW: 18226 47.96%


**Q9**: In at least one complete sentence, explain if your analysis of the samples accurately predicted the outcomes.

*(✏️ The supporter of kingfisher went beyond the upperbound of the sample, but was not far away from the upper bound of the 95% confidence interval, so this happens to be the 5% uncertainty of the interval. People who follows DUO falls within the 95% confidence interval and proves our samples predicted the outcomes. ✏️)*

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)